File: rwtest.cc
// c++-11 program to compute numerically the pdf of random walk displacments.
#include <iostream>
#include <iomanip>
#include <stdlib.h>
#include <random>
#include <ctime>

std::random_device rd;
std::mt19937 generator(rd()); // Mersenne Twister
std::bernoulli_distribution randbool(0.5);

int onewalk1(int num_coin_flips) {
  int num_heads = 0;
  for (int i=0; i < num_coin_flips; ++i)
    num_heads += randbool(generator);
  int num_tails = num_coin_flips - num_heads;
  return num_heads - num_tails;
}

int onewalk2(int num_coin_flips) {
  int num_heads = 0;
  for (int i=0; i < num_coin_flips; ++i)
    if(randbool(generator)) ++num_heads;
  int num_tails = num_coin_flips - num_heads;
  return num_heads - num_tails;
}

void run_walks (int num_trials, int num_steps, int * counts_on_lattice_sites,
                int (*walkfunction) (int), const char * walkfunction_name ) {
  std::clock_t c_start = std::clock();
  for(int i=0; i < num_trials; ++i) {
    // translate final displacement to be >= 0
    int final_displacement = (*walkfunction)(num_steps) + num_steps;
    counts_on_lattice_sites[final_displacement] += 1;
  }
  std::clock_t c_end = std::clock();
  double seconds =  (c_end-c_start)/ (double) CLOCKS_PER_SEC;
  std::cerr << "c++ " <<  walkfunction_name << ": " <<  std::fixed <<
    std::setprecision(2) << seconds << std::endl;
}

void print_walks(int num_trials, int num_steps, int * counts_on_lattice_sites, 
                 int num_sites_on_lattice) {
  for (int i=0; i < num_sites_on_lattice; ++i)
    if (counts_on_lattice_sites[i] > 0)
      std::cout << i - num_steps << " " << 
        (counts_on_lattice_sites[i] / (double) num_trials) << std::endl;
}

int main () {

  int num_steps = 5000;
  int num_sites_on_lattice = 2 * num_steps + 1;
  int num_trials = 100000;
  
  int * counts_on_lattice_sites = new int [num_sites_on_lattice];

  run_walks(num_trials, num_steps, counts_on_lattice_sites,onewalk1,"onewalk1");

  for(int i=0; i < num_sites_on_lattice; ++i)
    counts_on_lattice_sites[i] = 0;
  run_walks(num_trials, num_steps, counts_on_lattice_sites,onewalk2,"onewalk2");

  print_walks(num_trials, num_steps, counts_on_lattice_sites, num_sites_on_lattice);
  delete(counts_on_lattice_sites); 
  return 0;

}

File: rwtest.jl
#!/usr/bin/env julia

# Julia program to compute numerically the pdf of random walk displacments.

const num_steps = 3000
const num_trials = 100000

if VERSION < v"0.4.0-dev+980"
    macro ourrandbool()
        :(randbool())
    end
else
    macro ourrandbool()
        :(rand(Bool))
    end
end

function onewalk1(num_coin_flips)
    num_heads = 0
    for i in 1:num_coin_flips
        num_heads += @ourrandbool()
    end
    num_tails = num_coin_flips - num_heads
    return num_heads - num_tails
end

function onewalk2(num_coin_flips)
    num_heads = 0
    for i in 1:num_coin_flips
        @ourrandbool() ? num_heads += 1 : nothing
    end
    num_tails = num_coin_flips - num_heads
    return num_heads - num_tails
end

function run_walks (num_trials, num_steps, counts_on_lattice_sites, walkfunction,
                    walkfunctionname)
    time = @elapsed for i in 1:num_trials
        # translate final displacement to be >= 1
        final_displacement = walkfunction(num_steps)+num_steps+1  
        counts_on_lattice_sites[final_displacement] += 1
    end
    println(STDERR, "julia $walkfunctionname : $time")
end

function print_walks (num_trials, counts_on_lattice_sites)
    n = length(counts_on_lattice_sites)
    for i in 1:n
        if counts_on_lattice_sites[i] > 0
            println(i-(n+1)/2, " ", counts_on_lattice_sites[i]/num_trials)
        end
    end
end

# Run several walks and collect statistics
function run_simulation()
    num_sites_on_lattice = 2 * num_steps + 1
    counts_on_lattice_sites = zeros(Int,num_sites_on_lattice)
    run_walks(num_trials, num_steps, counts_on_lattice_sites,onewalk1,
                              "onewalk1")

# This is slower than the onewalk1    
#    fill!(counts_on_lattice_sites,0)
#    run_walks(num_trials, num_steps, counts_on_lattice_sites,onewalk2,
#                              "onewalk2")

    print_walks(num_trials, counts_on_lattice_sites)
end

# Run the program
run_simulation()
